### ISSUE TITLE
Automatically remove disk if instance failed to create

### ISSUE THREAD

**Issue by p0deje** [mitchellh/vagrant-google#112]
There is a corner case which may happen with the following steps:

1. Vagrantfile is configured to automatically create a disk.
2. User runs `vagrant up`.
3. vagrant-google creates disk.
4. vagrant-google tries to create instance, but something goes wrong (e.g. "Backend error").

Since instance is not created, disk should be properly removed. Subsequent attempts to create instance fail because disk with the same ID is already present.

```
/var/lib/jenkins/vagrant_home/gems/gems/fog-google-0.0.9/lib/fog/google.rb:232:in `build_excon_response': The resource 'projects/pr-1069/zones/us-central1-b/disks/example' already exists (Fog::Errors::Error)
```

I'm willing to fix the issue, but not sure how it should be properly handled.

**Comment by Temikus** [comment]
@p0deje This looks like a duplicate of #69 
I don't have time to get to it now, but patches are welcome and I'm willing to give pointers and review your code along the way.

The main thing to consider is that we never-ever want to remove an existing image if we didn't create it on vagrant up.
Vagrant-google allows to attach to existing images via disk-name, as per README:
> `disk_name` - The disk name to use. If the disk exists, it will be reused, otherwise created.

The relevant code is in `run_instance.rb` action:
```
 # Check if disk type is available in the zone and set the proper resource link
            disk_type = get_disk_type(env, disk_type, zone)

            if disk_name.nil?
              # no disk_name... disk_name defaults to instance name
              disk = env[:google_compute].disks.create(
                  name: name,
                  size_gb: disk_size,
                  type: disk_type,
                  zone_name: zone,
                  source_image: image
              )
              disk.wait_for { disk.ready? }
            else
              disk = env[:google_compute].disks.get(disk_name, zone)
              if disk.nil?
                # disk not found... create it with name
                disk = env[:google_compute].disks.create(
                    name: disk_name,
                    size_gb: disk_size,
                    type: disk_type,
                    zone_name: zone,
                    source_image: image
                )
                disk.wait_for { disk.ready? }
              end
            end
```
So, just deleting a disk in the zone with the same name as the instance doesn't seem like a good idea.
You'll need to devise a way to mark the drive as one created by vagrant and then insert the appropriate deletion logic.

If we're trying to catch issues on the step of the server being ready:
```
          begin
            server.wait_for { ready? }
            env[:metrics]["instance_ready_time"] = Time.now.to_i - request_start_time
            @logger.info("Time for instance ready: #{env[:metrics]["instance_ready_time"]}")
            env[:ui].info(I18n.t("vagrant_google.ready"))
          rescue
            env[:interrupted] = true
          end
```
, it may just be a question of passing the info about the disk deletion to another action (through `env` probably?) and adding the appropriate logic to `terminate_instance.rb` action, or making a custom action (something like `disk_cleanup.rb`) and adding it to the action builder in `action.rb`:
```
      # This action is called to terminate the remote machine.
      def self.action_destroy
        Vagrant::Action::Builder.new.tap do |b|
          b.use Call, DestroyConfirm do |env, b2|
            if env[:result]
              b2.use ConfigValidate
              b2.use Call, IsCreated do |env2, b3|
                unless env2[:result]
                  b3.use MessageNotCreated
                  next
                end
                b3.use ConnectGoogle
                b3.use TerminateInstance
              end
            else
              b2.use MessageWillNotDestroy
            end
          end
        end
      end
```

However, if we're trying to catch issues on the stage of `servers.create` (line 123-127 in `run_instance.rb`:
```
           :disks               => [disk.get_as_boot_disk(true, autodelete_disk)],
              :service_accounts    => service_accounts
            }
            server = env[:google_compute].servers.create(defaults)
            @logger.info("Machine '#{zone}:#{name}' created.")
```
, then you'll need to introduce proper termination handling in there. 

Let me know if you have any questions, happy to help :smile:

**Comment by p0deje** [comment]
I'd say the proper solution would be to catch Fog errors during instance creating, remove the disk and re-raise them.

I'll open a pull request tomorrow with an example.

**Comment by Temikus** [comment]
@p0deje Ack, will check it out.

2 things to keep in mind that you may run into:
- currently fog-google raises mostly RuntimeErrors, not proper exceptions.
- there are unfixed bugs in GCE API on Google's side (https://github.com/fog/fog-google/issues/45).

**Comment by p0deje** [comment]
In my particular case this was a "Backend Error":

```
$ timeout 10m vagrant up --provider=google
Bringing machine 'default' up with 'google' provider...
==> default: Launching an instance with the following settings...
==> default:  -- Name:            example
==> default:  -- Type:            n1-standard-4
==> default:  -- Disk type:       pd-ssd
==> default:  -- Disk size:       15 GB
==> default:  -- Disk name:       
==> default:  -- Image:           coreos-1445053959
==> default:  -- Zone:            us-central1-b
==> default:  -- Network:         default
==> default:  -- Metadata:        '{}'
==> default:  -- Tags:            '[]'
==> default:  -- IP Forward:      
==> default:  -- External IP:     
==> default:  -- Preemptible:     true
==> default:  -- Auto Restart:    false
==> default:  -- On Maintenance:  TERMINATE
==> default:  -- Autodelete Disk: true
==> default:  -- Scopes:          ["storage-ro"]
/var/lib/jenkins/vagrant_home/gems/gems/fog-google-0.0.9/lib/fog/google.rb:232:in `build_excon_response': Backend Error (Fog::Errors::Error)
	from /var/lib/jenkins/vagrant_home/gems/gems/fog-google-0.0.9/lib/fog/google.rb:203:in `request'
	from /var/lib/jenkins/vagrant_home/gems/gems/fog-google-0.0.9/lib/fog/google/requests/compute/get_disk.rb:47:in `get_disk'
	from /var/lib/jenkins/vagrant_home/gems/gems/fog-google-0.0.9/lib/fog/google/models/compute/disks.rb:25:in `get'
	from /var/lib/jenkins/vagrant_home/gems/gems/fog-google-0.0.9/lib/fog/google/models/compute/disk.rb:88:in `reload'
	from /var/lib/jenkins/vagrant_home/gems/gems/fog-core-1.32.1/lib/fog/core/model.rb:70:in `block in wait_for'
	from /var/lib/jenkins/vagrant_home/gems/gems/fog-core-1.32.1/lib/fog/core/wait_for.rb:7:in `block in wait_for'
	from /var/lib/jenkins/vagrant_home/gems/gems/fog-core-1.32.1/lib/fog/core/wait_for.rb:6:in `loop'
	from /var/lib/jenkins/vagrant_home/gems/gems/fog-core-1.32.1/lib/fog/core/wait_for.rb:6:in `wait_for'
	from /var/lib/jenkins/vagrant_home/gems/gems/fog-core-1.32.1/lib/fog/core/model.rb:69:in `wait_for'
	from /var/lib/jenkins/vagrant_home/gems/gems/vagrant-google-0.2.1/lib/vagrant-google/action/run_instance.rb:103:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/warden.rb:34:in `call'
	from /var/lib/jenkins/vagrant_home/gems/gems/vagrant-google-0.2.1/lib/vagrant-google/action/warn_networks.rb:28:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/warden.rb:34:in `call'
	from /var/lib/jenkins/vagrant_home/gems/gems/vagrant-google-0.2.1/lib/vagrant-google/action/sync_folders.rb:33:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/warden.rb:34:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/builtin/provision.rb:80:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/warden.rb:34:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/warden.rb:95:in `block in finalize_action'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/warden.rb:34:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/warden.rb:34:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/builder.rb:116:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/runner.rb:66:in `block in run'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/util/busy.rb:19:in `busy'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/runner.rb:66:in `run'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/builtin/call.rb:53:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/warden.rb:34:in `call'
	from /var/lib/jenkins/vagrant_home/gems/gems/vagrant-google-0.2.1/lib/vagrant-google/action/connect_google.rb:47:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/warden.rb:34:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/builtin/box_check_outdated.rb:36:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/warden.rb:34:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/builtin/config_validate.rb:25:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/warden.rb:34:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/builtin/handle_box.rb:56:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/warden.rb:34:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/builder.rb:116:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/runner.rb:66:in `block in run'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/util/busy.rb:19:in `busy'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/action/runner.rb:66:in `run'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/machine.rb:214:in `action_raw'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/machine.rb:191:in `block in action'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/environment.rb:516:in `lock'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/machine.rb:178:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/machine.rb:178:in `action'
	from /opt/vagrant/embedded/gems/gems/vagrant-1.7.3/lib/vagrant/batch_action.rb:82:in `block (2 levels) in run'

```

**Comment by p0deje** [comment]
@Temikus I've submitted #113, please take a look and let me know what you think.

**Comment by Temikus** [comment]
Fixed by #113, thanks again @p0deje!

@erjohnso - can you please mark this as fixed?
